### ISSUE TITLE
Continuous integration testing

### ISSUE THREAD

**Issue by welwood08** [DNSCrypt/dnscrypt-proxy#1057]
It has become increasingly apparent that something is needed to catch obvious problems before a release! I'm not pointing fingers, I just mean that testing is tedious and difficult to get right. I've been fiddling with various options in the background for a while but now that another release has had a critical problem, it's probably a good idea to have a place to discuss this need.

The current Travis setup is slow and doesn't really test anything. Adding unit tests to it is an easy win, but not much of the code is covered by unit tests yet and they wouldn't have caught either of the recent panic problems. On the assumption that integration tests would have a non-trivial runtime cost, I decided I first needed to speed up the existing automation.

The most significant runtime improvement can be achieved by cross-compiling in parallel, but Travis hits a limit because the Android NDK takes a while to get ready. I tried to stick with the project's existing CI setup but after fighting with the syntax, once I hit that limit I expanded my search and discovered GitHub Actions had just come out of beta.

<table><thead><tr><th/><th><a href="https://github.com/welwood08/dnscrypt-proxy/compare/master...travis-experiments">stay with Travis-CI</a></th><th><a href="https://github.com/welwood08/dnscrypt-proxy/compare/master...ghactions-experiments">switch to GitHub Actions</a></th></tr></thead><tbody>
<tr><td>Current runtime</td><td>~13mins</td><td>N/A</td></tr>
<tr><td>Best runtime found</td><td>~7mins</td><td>~3mins</td></tr>
<tr><td>Pros</td><td><ul><li>Already set up</li><li>Recently added multiarchitecture support with real hardware</li><li>GitHub doesn't know the signing key</li></ul></td><td><ul><li>Build artifacts for every run hosted and available to public</li><li>Latest Android NDK preinstalled</li><li>Secrets can be exposed granularly to specific steps that need them</li></ul></td></tr>
<tr><td>Cons</td><td><ul><li>Syntax feels like features have been bolted on</li><li>MacOS machines slow to start</li></ul></td><td><ul><li>Syntax is much more verbose</li><li>Only out of beta very recently</li><li>UI is clunky and takes a while to realise a run is finished</li></ul></td></tr>
</tbody></table>

When I realised build artifacts were public on Actions, I decided it'd be cool to always sign them. I didn't realise until I was composing this that it could be significant for the signing key to remain unknown to the service hosting the releases. If that's the case, we could continue to use Travis just for releases and use Actions as faster PR and commit feedback. As mentioned in https://github.com/DNSCrypt/dnscrypt-proxy/pull/989#issuecomment-548971186 I believe @jedisct1 is already on board with using multiple services and making CI faster, but should I leave Travis active for signing releases and not do any signing in Actions?

TLDR: I'm working on adding integration tests!

**Comment by jedisct1** [comment]
Having integration tests would indeed be very useful!

One thing Travis does that Github Actions don’t is support ARM architectures. Travis recently added ARM hardware, and given that many dnscrypt-proxy users use it on ARM router and Raspberry Pi, it can be interesting to run tests on these machines.

Other than that, if Github Actions are faster, have less limits, can do everything Travis can, plus run the integration tests, using them sounds like an excellent idea :)

Having the artifacts signed by a different service doesn’t really improve security, as the source code comes from Github anyway.

I recently switched to Azure Pipelines for another project (libsodium) and was pleasantly surprised. It’s fast, and the UI is really good, including to build the pipelines. But if you already have something using Actions, let’s stick to that!

**Comment by MindTooth** [comment]
https://blog.travis-ci.com/2019-11-12-multi-cpu-architecture-ibm-power-ibm-z

Travis now even supports IBM architectures.  Not sure that is something that you develop towards, but it's there. 👍

**Comment by welwood08** [comment]
I did mention Travis's multi-arch support as a pro in the table above if real hardware is important, but also found an Action that can do ARM testing via qemu and docker if it's not.

For IBM architectures in particular, official release binaries aren't currently produced and that's presumably because no demand exists for official binaries. I personally wouldn't want to commit to supporting something nobody was using, but I'm sure if it was requested it would be considered.

I'm making the integration tests standalone so I can iterate locally, but that also has the added bonus that they should work on any CI service so we don't necessarily have to pick only one. Pipelines was on my shortlist but since Microsoft owns Azure and GitHub, and Actions mostly runs on the same VM infrastructure I picked the one I already had an account for - if it turns out to be significantly different I'm open to looking at it more closely, as well as any other suggestions anybody has.

**Comment by jedisct1** [comment]
IBM Z is nice, but indeed not something we should really care much about since no one ever asked for IBM Z binaries. 

Let's stick to Github Actions. ARM testing via qemu is fine.
